docs/design/sql/16-nested-fields-access.md

```diff
@@ -0,0 +1,359 @@
+case-sensitive. This is contrary to built-in types, whose names are
+case-insensitive. This might seem illogical, but is consistent with the rest of
+our SQL implementation where identifiers are case-sensitive, but keywords (which
+include built-in functions and types) are not.
+
+```sql
+CREATE TYPE My_Type ...;
+
+CREATE MAPPING m (
+  field my_type  -- fails here
+) ...
+```
+
+Here we need to note that identifiers in SQL are always case-sensitive, however,
+many SQL implementations convert unquoted identifiers to canonical case, which
+gives the impression that identifiers are case-insensitive. Quoted identifiers
+are always case-sensitive. The conversion to canonical case is optional in the
+SQL standard, Hazelcast doesn't do it for legacy reasons.
+
+## Automatic resolution
+
+When a mapping or another type is auto-resolved, for non-primitive types we
+check the SQL catalog if a type exists.
+
+## UDTs and the top-level fields
+
+Currently, if the top-level `__key` or `this` field of an imap isn't a SQL
+primitive type, we make the field available as a hidden field, and all the
+fields are expanded as top-level fields. For example, if we have `IMap<Long,
+Person>`, then the mapping will have these fields:
+
+- `__key BIGINT`
+- `this OBJECT` (hidden)
+- `name`, `birthDate`, ... (all fields of the `Person` class)
+
+This behavior remains unchanged, that is, even if there is a UDT for `Person`,
+`this` will have `OBJECT` type. 
+
+## ROW values
+
+The SQL standards specifies the ROW type. A ROW value contains an array of
+values, but doesn't contain field names. It is important to note that UDT values
+aren't ROW values. However, they can be converted both ways:
+
+- `TO_ROW` function converts an UDT value to `ROW` value
+
+- `CAST(rowExpression AS <UDT name>)`: converts a `ROW` value to a UDT value.
+  The ROW value must have the same number and types of fields as the UDT, in the
+  same order.
+
+When the ROW value is returned to the client, a
+`com.hazelcast.sql.impl.expression.RowValue` is returned to the java client.
+Note that it is a private API class - the reason is that STRUCT types aren't
+supported by JDBC. We could make this API public in our custom SQL API, but we
+consider JDBC to be the primary Java API. This is also what PG and Oracle do
+(TODO confirm Oracle). The user can reasonably only use `toString` on the value,
+which is what JDBC UI clients do, the `toString` will contain comma-separated
+list of all fields.
+
+The `ROW` type supports comparison operators that is different from how the
+comparison operator works with `OBJECT` type. It compares all field one by one,
+from left to right (this is important for operators other than `=`).
+
+The `ROW` type can also be assigned to another `ROW` type, provided they have
+the same number and types of fields. This is important with `ROW` literals, they
+can be used to create UDT values:
+
+```sql
+UPDATE m
+SET address=CAST(('street', 'city', 12345, 'Slovakia') AS AddressType
```

Comment:
Currently this CAST is "implicit" (or rather it doesn't happen, the UpsertTarget does all the work for this).